### ISSUE TITLE
Consistency check possible removal of legitime flow

### ISSUE THREAD

**Issue by italovalcy** [kytos-ng/flow_manager#69]
I was running a few tests in disruptive scenarios (i.e., scenarios with link failures/flaps) and with a bunch of EVCs (around 100 EVCs) and everything worked as expected, except for one test in which one EVC didn't work as expected (after the failed link return to its normal state - UP - the end user started seen packet loss).

The test consists of the following workflow:

1. 100 EVCs are created using dynamic path
2. After the creation of the EVCs, we start a PING connectivity test between the endpoints of each EVC
3. A few seconds later, the link `00:00:00:00:00:00:00:01:11--00:00:00:00:00:00:00:06:11` goes down
4. 60 seconds later, the link `00:00:00:00:00:00:00:01:11--00:00:00:00:00:00:00:06:11` come up
5. ping keeps running for more 30 seconds and the test finishes.

Due to a problem in the way link down events are being handled by mef_eline, the EVC does not actually gets moved to the backup path during the link failure. However, after the link has come UP again, mef_eline runs a reprovisioning on the EVC and setup new flows (actually, at this point mef_eline was still running the loop over the EVCs affected by the link failure.. so that, the reprovisioning of the EVC will be through the same path, but now using different service vlans).

Below is a sequence of filtered log messages showing the behavior in terms of mef_eline (evc creation and reprovisioning), of_core (interface down and up) and flow_manager (flows creation, removal and consistency check):

```
# zcat /var/log/kytos/kytos.log-20220203012301.gz | egrep "12319557280506940481|f7dc1326b18c41|Consistency check:|Modified"
##
## Note: below is the EVC creation
##
2022-02-03 06:27:16,977 - INFO [kytos.napps.kytos/flow_manager] (Thread-916) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:27:17,202 - INFO [kytos.napps.kytos/flow_manager] (Thread-917) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:27:17,259 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:27:17,418 - INFO [kytos.napps.kytos/flow_manager] (Thread-919) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 16, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 859}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 859}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 16}]}], 'force': False}
2022-02-03 06:27:17,870 - INFO [kytos.napps.kytos/flow_manager] (Thread-920) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 21, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 859}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 859}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 21}]}], 'force': False}
2022-02-03 06:27:18,451 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:27:18,466 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC(f7dc1326b18c41, my evc 1149) was deployed.
##
## Note: Link goes down
##
2022-02-03 06:28:51,760 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_31) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:01')) 00:00:00:00:00:00:00:01:11
2022-02-03 06:28:51,821 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_1) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:06')) 00:00:00:00:00:00:00:06:11
...
##
## Note: Link come up again
##
2022-02-03 06:29:55,311 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_8) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:06')) 00:00:00:00:00:00:00:06:11
2022-02-03 06:29:55,416 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_21) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:01')) 00:00:00:00:00:00:00:01:11
##
## Note: mef_eline path  reprovision for the EVC
##
2022-02-03 06:30:38,073 - INFO [kytos.napps.kytos/flow_manager] (Thread-1265) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:30:38,298 - INFO [kytos.napps.kytos/flow_manager] (Thread-1266) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:30:38,591 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:30:38,779 - INFO [kytos.napps.kytos/flow_manager] (Thread-1268) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 16, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 1455}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 1455}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 16}]}], 'force': False}
2022-02-03 06:30:39,071 - INFO [kytos.napps.kytos/flow_manager] (Thread-1269) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 21, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 1455}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 1455}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 21}]}], 'force': False}
2022-02-03 06:30:39,964 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:30:40,122 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC(f7dc1326b18c41, my evc 1149) was deployed.
##
## Note: flow_manager consistency routine
##
2022-02-03 06:30:41,051 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,108 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Flow forwarded to switch 00:00:00:00:00:00:00:06 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:06', 'table_id': 0, 'match': {'in_port': 21, 'dl_vlan': 1149}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319557280506940481, 'id': '13ef7612927e97f13d3f1942ff3cdf71', 'stats': {'byte_count': 320166, 'duration_sec': 198, 'duration_nsec': 426511681, 'packet_count': 2541}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'vlan_id': 859, 'action_type': 'set_vlan'}, {'port': 11, 'action_type': 'output'}]}]}]}
2022-02-03 06:30:41,146 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,305 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Flow forwarded to switch 00:00:00:00:00:00:00:06 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:06', 'table_id': 0, 'match': {'in_port': 11, 'dl_vlan': 859}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319557280506940481, 'id': '9dc8c461eb1e6befea3cc50e303f8d58', 'stats': {'byte_count': 320166, 'duration_sec': 198, 'duration_nsec': 369709443, 'packet_count': 2541}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'vlan_id': 1149, 'action_type': 'set_vlan'}, {'port': 21, 'action_type': 'output'}]}]}]}
2022-02-03 06:30:41,349 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,460 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:43,314 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: missing flow on switch 00:00:00:00:00:00:00:06.
```

The above log messages show the situation described before. It is important to notice the last set of messages: the ones related to the "flow_manager consistency routine". As discussed before, mef_eline takes longer to react to the link failure for this EVC, and it ends up using the same (recovered) path but with different service-vlans. The customer VLAN is 1149, and previous service VLAN was 859 while the new service VLAN is  1455. Everything should be working at the time `2022-02-03 06:30:40,122` (when the evc was reprovisioned). However, at `2022-02-03 06:30:41,051 ` the flow_manager consistency routine identifies alien flows at the switch 00:00:00:00:00:00:00:06 and sends out a FlowMod Remove. The problem is at the first flow removal, where the match conditions are the same as the correct flow at the switch. Unfortunately, the there is no evidence in the logs about the consistency check fixing this mistake. However, a dump of the switch's flow_table at the end of the test reveals that indeed switch 00:06 have one flow missing:
```
$ grep -A 10 f7dc1326b18c41 dump-flowtable-20220203012301-xxx.xxx.xxx.229
        Cookie           = aaf7dc1326b18c41
        Send_flow_rem    = true
        Persistent       = false
        [MATCHFIELDS]
            OFPXMT_OFB_IN_PORT = 11
            OFPXMT_OFB_VLAN_VID = 1455(OFPVID_PRESENT=1)
        [INSTRUCTIONS]
            [OFPIT_APPLY_ACTIONS]
                 [ACTIONS]
                    [OFPAT_POP_VLAN]
                    [OFPAT_PUSH_VLAN]
```

I've tried to extract the flow.persistency objects from storehouse for this execution, however due to another issue (#67) it was not readable.

**Comment by viniarck** [comment]
@italovalcy I'm suspecting that this might have been a side effect of storehouse being corrupted, and the [`_add_flow_store`](https://github.com/kytos-ng/flow_manager/blob/master/storehouse.py#L120-L125) method when calling `storehouse.save_flow` event crashed and didn't even return an error? and it also has some shared data being passed around, not sure which value that could end up. For instance, in the log above we don't see any log messages from the `save_flow_callback` method such as:

```
[kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_1) Flow saved in kytos.flow.persistence.f096a6aae17048afbd38d1914b492879
```

or:

```
log.error(f"Can't update persistence box {data.box_id}.")
```

With that being said, it makes sense that it would consider an alien and delete it since what was being written to the storehouse wasn't functioning properly, although having this silent sort of crash also sucks. It would be cool to try to reproduce this again once #67 lands and decide what to do and prioritize next, thanks for reporting this.

**Comment by italovalcy** [comment]
@viniarck, thanks for the analysis and the suggestions! Yes, it does make total sense, and I agree with you. I wanna add an extra comment regarding the detection of alien flows: since there was a "recent" flow removal by cookie-id (see logs at `2022-02-03 06:30:38,073 ` and `2022-02-03 06:30:38,298`), the consistency check routine could benefit from that and avoid sending a new flow_removal for the "alien flow". That way, the EVC will remain consistent in the flow table. Without this optimization, we will see a flow_removal from the alien detection followed by a flow add from the missing flow detection.

I was able to reproduce this behavior with the following test:

1. Run Kytos-ng with docker image `amlight/kytos:latest`
2. Run Mininet with a ring topology:
```
# cat ring.py
from mininet.topo import Topo

class RingTopo(Topo):
    """Ring topology with three switches
    and one host connected to each switch"""

    def build(self):
        # Create two hosts
        h1 = self.addHost('h1', ip='0.0.0.0')
        h2 = self.addHost('h2', ip='0.0.0.0')
        h3 = self.addHost('h3', ip='0.0.0.0')

        # Create the switches
        s1 = self.addSwitch('s1')
        s2 = self.addSwitch('s2')
        s3 = self.addSwitch('s3')

        # Add links between the switch and each host
        self.addLink(s1, h1)
        self.addLink(s2, h2)
        self.addLink(s3, h3)

        # Add links between the switches
        self.addLink(s1, s2)
        self.addLink(s2, s3)
        self.addLink(s3, s1)

topos = {
    'ring': (lambda: RingTopo()),
}

# mn --custom ring.py --topo ring --controller=remote,ip=127.0.0.1
```
2. Change flow_stats interval in `/src/kytos-of-core/settings.py` with `STATS_INTERVAL = 15` and restart kytos
3. Create a few EVCs (I've created 100 EVCs) between h1 and h2:
```
for vlan in $(seq 401 500); do curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "evc-vlan-'$vlan'", "dynamic_backup_path": true, "enabled": true, "uni_z": {"tag": {"value": '$vlan', "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_a": {"tag": {"value": '$vlan', "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:02:1"}}'; done
```
4. Wait a few seconds to get the flows created and disable the link s1--s2, to force the reprovisioning of the EVCs (now through s1-s3-s2):
```
ip link set down dev s1-eth2 && sleep 30 && ip link set up dev s1-eth2
```
5. Monitor the EVCs to make sure they are operational (i.e., they have the proper number of flows). I used the following script:
```
# s=s1; for vlan in $(seq 401 500); do if [ $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1" | wc -l) -ne 2 ]; then echo missing flows for $vlan on switch $s -- current flows = $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1"); fi; done

# s=s2; for vlan in $(seq 401 500); do if [ $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1" | wc -l) -ne 2 ]; then echo missing flows for $vlan on switch $s -- current flows = $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1"); fi; done
missing flows for 439 on switch s2 -- current flows   = cookie=0xaaac2133b837bb4d, duration=56.868s, table=0, n_packets=0, n_bytes=0, idle_age=56, in_port=2,dl_vlan=1975 actions=strip_vlan,mod_vlan_vid:439,output:1
```

The above scenario may not happen in the first execution. In fact, I had to repeat the above test a few times to get into that situation (reducing the FLOW_STATS interval helped).

The situation shown above was caused by a unnecessary flow_removal sent by flow_manager as part of the consistency check (alien flow detected). After a few seconds, the missing flow will be recreated, as part of the consistency routine. Maybe we could optimize the above situation by avoiding send flow_removal for "recently removed flows" (whey they match each other, of course).

The logs for the above situation are shown below:
```
## Notice: first I filter by the time interval in which the error happened (this is the file "last-event"
# egrep "^Feb  4 21:03:0[1-9]|^Feb  4 21:03:[1-9]|^Feb  4 21:04:" /var/log/syslog > last-event
# egrep "12319564338740338244|f7e27e86301e44|Consistency" last-event
Feb  4 21:03:01 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:02 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:03 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:03 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:16 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:16 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:18 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
Feb  4 21:03:33 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:33 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:46 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:46 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
Feb  4 21:03:47 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319564338740338244, 'cookie_mask': 18446744073709551615}], 'force': True}
Feb  4 21:03:47 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319564338740338244, 'cookie_mask': 18446744073709551615}], 'force': True}
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:141:  EVC f7e27e86301e44 was synced to the storehouse.
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}, {'match': {'in_port': 2, 'dl_vlan': 3760}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 483}, {'action_type': 'output', 'port': 1}]}], 'force': False}
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:462:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:02', 'table_id': 0, 'match': {'in_port': 1, 'dl_vlan': 483}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319564338740338244, 'id': 'c801fc231c4405a934be36fc2345798c', 'stats': {'byte_count': 0, 'duration_sec': 156, 'duration_nsec': 21000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'vlan_id': 2048, 'action_type': 'set_vlan'}, {'port': 2, 'action_type': 'output'}]}]}]}
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:462:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:02', 'table_id': 0, 'match': {'in_port': 2, 'dl_vlan': 2048}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319564338740338244, 'id': 'fbeb5f41489e70fb990a8504b3035dc2', 'stats': {'byte_count': 0, 'duration_sec': 155, 'duration_nsec': 962000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'vlan_id': 483, 'action_type': 'set_vlan'}, {'port': 1, 'action_type': 'output'}]}]}]}
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}, {'match': {'in_port': 2, 'dl_vlan': 3760}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 483}, {'action_type': 'output', 'port': 1}]}], 'force': False}
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:141:  EVC f7e27e86301e44 was synced to the storehouse.
Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:552:  EVC(f7e27e86301e44, evc-vlan-483) was deployed.
Feb  4 21:04:17 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
Feb  4 21:04:17 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:405:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be installed. Flow: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}]}
```

**Comment by viniarck** [comment]
> @viniarck, thanks for the analysis and the suggestions! Yes, it does make total sense, and I agree with you. I wanna add an extra comment regarding the detection of alien flows: since there was a "recent" flow removal by cookie-id (see logs at `2022-02-03 06:30:38,073 ` and `2022-02-03 06:30:38,298`), the consistency check routine could benefit from that and avoid sending a new flow_removal for the "alien flow". That way, the EVC will remain consistent in the flow table. Without this optimization, we will see a flow_removal from the alien detection followed by a flow add from the missing flow detection.
> 
> I was able to reproduce this behavior with the following test:
> 
> 1. Run Kytos-ng with docker image `amlight/kytos:latest`
> 2. Run Mininet with a ring topology:
> 
> ```
> # cat ring.py
> from mininet.topo import Topo
> 
> class RingTopo(Topo):
>     """Ring topology with three switches
>     and one host connected to each switch"""
> 
>     def build(self):
>         # Create two hosts
>         h1 = self.addHost('h1', ip='0.0.0.0')
>         h2 = self.addHost('h2', ip='0.0.0.0')
>         h3 = self.addHost('h3', ip='0.0.0.0')
> 
>         # Create the switches
>         s1 = self.addSwitch('s1')
>         s2 = self.addSwitch('s2')
>         s3 = self.addSwitch('s3')
> 
>         # Add links between the switch and each host
>         self.addLink(s1, h1)
>         self.addLink(s2, h2)
>         self.addLink(s3, h3)
> 
>         # Add links between the switches
>         self.addLink(s1, s2)
>         self.addLink(s2, s3)
>         self.addLink(s3, s1)
> 
> topos = {
>     'ring': (lambda: RingTopo()),
> }
> 
> # mn --custom ring.py --topo ring --controller=remote,ip=127.0.0.1
> ```
> 
> 2. Change flow_stats interval in `/src/kytos-of-core/settings.py` with `STATS_INTERVAL = 15` and restart kytos
> 3. Create a few EVCs (I've created 100 EVCs) between h1 and h2:
> 
> ```
> for vlan in $(seq 401 500); do curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "evc-vlan-'$vlan'", "dynamic_backup_path": true, "enabled": true, "uni_z": {"tag": {"value": '$vlan', "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_a": {"tag": {"value": '$vlan', "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:02:1"}}'; done
> ```
> 
> 4. Wait a few seconds to get the flows created and disable the link s1--s2, to force the reprovisioning of the EVCs (now through s1-s3-s2):
> 
> ```
> ip link set down dev s1-eth2 && sleep 30 && ip link set up dev s1-eth2
> ```
> 
> 5. Monitor the EVCs to make sure they are operational (i.e., they have the proper number of flows). I used the following script:
> 
> ```
> # s=s1; for vlan in $(seq 401 500); do if [ $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1" | wc -l) -ne 2 ]; then echo missing flows for $vlan on switch $s -- current flows = $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1"); fi; done
> 
> # s=s2; for vlan in $(seq 401 500); do if [ $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1" | wc -l) -ne 2 ]; then echo missing flows for $vlan on switch $s -- current flows = $(ovs-ofctl dump-flows $s | egrep "in_port=1,dl_vlan=$vlan |vlan_vid:$vlan,output:1"); fi; done
> missing flows for 439 on switch s2 -- current flows   = cookie=0xaaac2133b837bb4d, duration=56.868s, table=0, n_packets=0, n_bytes=0, idle_age=56, in_port=2,dl_vlan=1975 actions=strip_vlan,mod_vlan_vid:439,output:1
> ```
> 
> The above scenario may not happen in the first execution. In fact, I had to repeat the above test a few times to get into that situation (reducing the FLOW_STATS interval helped).
> 
> The situation shown above was caused by a unnecessary flow_removal sent by flow_manager as part of the consistency check (alien flow detected). After a few seconds, the missing flow will be recreated, as part of the consistency routine. Maybe we could optimize the above situation by avoiding send flow_removal for "recently removed flows" (whey they match each other, of course).
> 
> The logs for the above situation are shown below:
> 
> ```
> ## Notice: first I filter by the time interval in which the error happened (this is the file "last-event"
> # egrep "^Feb  4 21:03:0[1-9]|^Feb  4 21:03:[1-9]|^Feb  4 21:04:" /var/log/syslog > last-event
> # egrep "12319564338740338244|f7e27e86301e44|Consistency" last-event
> Feb  4 21:03:01 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:02 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:03 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:03 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:16 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:16 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:18 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
> Feb  4 21:03:33 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:33 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:46 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:46 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
> Feb  4 21:03:47 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319564338740338244, 'cookie_mask': 18446744073709551615}], 'force': True}
> Feb  4 21:03:47 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319564338740338244, 'cookie_mask': 18446744073709551615}], 'force': True}
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:141:  EVC f7e27e86301e44 was synced to the storehouse.
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}, {'match': {'in_port': 2, 'dl_vlan': 3760}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 483}, {'action_type': 'output', 'port': 1}]}], 'force': False}
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:462:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:02', 'table_id': 0, 'match': {'in_port': 1, 'dl_vlan': 483}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319564338740338244, 'id': 'c801fc231c4405a934be36fc2345798c', 'stats': {'byte_count': 0, 'duration_sec': 156, 'duration_nsec': 21000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'vlan_id': 2048, 'action_type': 'set_vlan'}, {'port': 2, 'action_type': 'output'}]}]}]}
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:457:  Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:462:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:02', 'table_id': 0, 'match': {'in_port': 2, 'dl_vlan': 2048}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319564338740338244, 'id': 'fbeb5f41489e70fb990a8504b3035dc2', 'stats': {'byte_count': 0, 'duration_sec': 155, 'duration_nsec': 962000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'vlan_id': 483, 'action_type': 'set_vlan'}, {'port': 1, 'action_type': 'output'}]}]}]}
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:737:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}, {'match': {'in_port': 2, 'dl_vlan': 3760}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 483}, {'action_type': 'output', 'port': 1}]}], 'force': False}
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:141:  EVC f7e27e86301e44 was synced to the storehouse.
> Feb  4 21:03:48 012f88145fe6 kytos.napps.kytos/mef_eline:INFO evc:552:  EVC(f7e27e86301e44, evc-vlan-483) was deployed.
> Feb  4 21:04:17 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:401:  Consistency check: missing flow on switch 00:00:00:00:00:00:00:02.
> Feb  4 21:04:17 012f88145fe6 kytos.napps.kytos/flow_manager:INFO main:405:  Flow forwarded to switch 00:00:00:00:00:00:00:02 to be installed. Flow: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 483}, 'cookie': 12319564338740338244, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3760}, {'action_type': 'output', 'port': 2}]}]}
> ```

Got it, @italovalcy, great idea, let's go for it. Thanks for the scripts and topology there were super helpful. I've prioritized and  opened this PR, I'm also gonna run the end to end tests with this PR suite shortly.
